[SPARK-8542][MLlib]PMML export for Decision Trees

JasmineGeorge opened the pull request:
Added support for PMML export for Decision Trees

Commit:
added Decisiontree PMML Exporter first cut

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
updated code for handling DT classifier, and categorical features

Commit:
updated code for handling DT classifier, and categorical features

Commit:
added unit tests

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
added more unittests and comments

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
style check changes

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

mengxr commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39291 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39291/consoleFull) for   PR 7842 at commit [`bae58fa`](https://github.com/apache/spark/commit/bae58fa8f0901b5fe92b71ac320dd642e9c9f4ac).

SparkQA commented:
[Test build #39291 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39291/console) for   PR 7842 at commit [`bae58fa`](https://github.com/apache/spark/commit/bae58fa8f0901b5fe92b71ac320dd642e9c9f4ac).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DecisionTreeModel(val topNode: Node, val algo: Algo) extends Serializable with Saveable with PMMLExportable`

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
style check changes

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39304 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39304/consoleFull) for   PR 7842 at commit [`6acc424`](https://github.com/apache/spark/commit/6acc4241569dc834910d1fa0be0f8b4de5d6f923).

SparkQA commented:
[Test build #39304 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39304/console) for   PR 7842 at commit [`6acc424`](https://github.com/apache/spark/commit/6acc4241569dc834910d1fa0be0f8b4de5d6f923).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DecisionTreeModel(val topNode: Node, val algo: Algo) extends Serializable`
  * `case class SoundEx(child: Expression) extends UnaryExpression with ExpectsInputTypes `

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
fixed style check failures

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39305 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39305/consoleFull) for   PR 7842 at commit [`f04404f`](https://github.com/apache/spark/commit/f04404fe93faefb84deec441df6aa1ce65965728).

SparkQA commented:
[Test build #39305 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39305/console) for   PR 7842 at commit [`f04404f`](https://github.com/apache/spark/commit/f04404fe93faefb84deec441df6aa1ce65965728).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DecisionTreeModel(val topNode: Node, val algo: Algo) extends Serializable`

AmplabJenkins commented:
Merged build finished. Test PASSed.

mengxr commented:
@selvinsource @coderxiang Do you have time to review this PR? Thanks!

coderxiang commented:
@mengxr looking

coderxiang commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExport.scala:
```diff
@@ -0,0 +1,49 @@
+import org.apache.spark.mllib.tree.model.DecisionTreeModel
+import org.dmg.pmml.DataDictionary
+
+import scala.collection.JavaConverters._
```
check order of imports, https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports

JasmineGeorge replied:
Let me fix this.

coderxiang commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/TreeModelUtils.scala:
```diff
@@ -0,0 +1,274 @@
+private[mllib] object TreeModelUtils {
+
+
+  def getPMMLTree(mllibTreeModel: DecisionTreeModel, modelName: String): TreeModel = {
```
is `fromPMMLTree` a better name here?

JasmineGeorge replied:
fromPMMLTree sounds good. Will make the change.

JasmineGeorge replied:
Actually this method takes MLLib tree and returns a PMML tree. So should we call it toPMMLTree ?

coderxiang replied:
sounds good!

JasmineGeorge replied:
Changes done !

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
organized import and function getPMMLTree renamed to toPMMLTree

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40337 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40337/consoleFull) for   PR 7842 at commit [`72ac3fa`](https://github.com/apache/spark/commit/72ac3faed527c0daaca93c7e165384ac19e5ad7c).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40339/consoleFull) for   PR 7842 at commit [`f151493`](https://github.com/apache/spark/commit/f151493b5043f1532ab4fbde8eaffa5c673d7260).

SparkQA commented:
[Test build #40337 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40337/console) for   PR 7842 at commit [`72ac3fa`](https://github.com/apache/spark/commit/72ac3faed527c0daaca93c7e165384ac19e5ad7c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #40339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40339/console) for   PR 7842 at commit [`f151493`](https://github.com/apache/spark/commit/f151493b5043f1532ab4fbde8eaffa5c673d7260).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JasmineGeorge commented:
changes requested are done !

coderxiang commented:
@JasmineGeorge Can you check the imports order? There are still some violations.

coderxiang commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExport.scala:
```diff
@@ -0,0 +1,50 @@
+    val dataFields = TreeModelUtils.getDataFieldsForTree(model)
+
+    val dataDictionary = new DataDictionary()
+    .withDataFields(dataFields.asJava)
```
use two space indent here

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

JasmineGeorge committed:
fixed imports and indentation

JasmineGeorge committed:
fixed imports and indentation

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41135/consoleFull) for   PR 7842 at commit [`c7799cb`](https://github.com/apache/spark/commit/c7799cb9d64fb75b0c9e2176f3f2f6a480f892f0).

SparkQA commented:
[Test build #41135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41135/console) for   PR 7842 at commit [`c7799cb`](https://github.com/apache/spark/commit/c7799cb9d64fb75b0c9e2176f3f2f6a480f892f0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DecisionTreeModel(val topNode: Node, val algo: Algo) extends Serializable`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41135/
Test PASSed.

JasmineGeorge commented:
I got the scala import formatter plugin to work finally. This time the imports should be formatted correctly.

coderxiang commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLModelExportFactory.scala:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.mllib.pmml.export
 
+import org.apache.spark.mllib.tree.model.DecisionTreeModel
```
here?

coderxiang commented on mllib/src/main/scala/org/apache/spark/mllib/tree/model/DecisionTreeModel.scala:
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.mllib.tree.model
 
+import org.apache.spark.mllib.pmml.PMMLExportable
```
ditto

JasmineGeorge committed:
more import formatting

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41145 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41145/consoleFull) for   PR 7842 at commit [`8b31f7d`](https://github.com/apache/spark/commit/8b31f7dd1db2f453a0373fc2b58c2fbc972fdad0).

SparkQA commented:
[Test build #41145 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41145/console) for   PR 7842 at commit [`8b31f7d`](https://github.com/apache/spark/commit/8b31f7dd1db2f453a0373fc2b58c2fbc972fdad0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DecisionTreeModel(val topNode: Node, val algo: Algo) extends Serializable`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41145/
Test PASSed.

JasmineGeorge commented:
formatted those files as well. Sorry I missed them last time.

selvinsource commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLModelExportFactory.scala:
```diff
@@ -55,6 +53,7 @@ private[mllib] object PMMLModelExportFactory {
           throw new IllegalArgumentException(
             "PMML Export not supported for Multinomial Logistic Regression")
         }
+      case decsionTreeModel : DecisionTreeModel => new DecisionTreePMMLModelExport(decsionTreeModel)
```
Spelling mistake: change decsionTreeModel to decisionTreeModel

selvinsource commented on mllib/src/test/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExportSuite.scala:
```diff
@@ -0,0 +1,364 @@
+
+    // get the pmml exporter for the DT and verify its the right exporter
+    val pmmlExporterForDT = PMMLModelExportFactory.createPMMLModelExport(decisionTreeModel)
+    assert(pmmlExporterForDT.isInstanceOf[DecisionTreePMMLModelExport])
```
This test on the right exporter could be moved to PMMLModelExportFactorySuite where I kept all the exporter checks for the factory.

selvinsource commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/TreeModelUtils.scala:
```diff
@@ -0,0 +1,275 @@
+      }
+    }
+
+    val miningFields = getMiningFieldsForTreeRec(List(treeModel.topNode), List[MiningField]())
```
getMiningFieldsForTreeRec is very similar to getDataFieldsForTreeRec, couldn't you merge these?
Also, I am not sure you even need getMiningFieldsForTreeRec, you could create the mining fields from the data fields you already have, usually (in the models I exported) the field names were the same for mining fields and data fields.

selvinsource commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/TreeModelUtils.scala:
```diff
@@ -0,0 +1,275 @@
+    if (node.split.isDefined) {
+      val split = node.split.get
+      val featureType = split.featureType
+      val fieldName = FieldName.create("field_" + split.feature.toString)
```
Here you are recreating the Field Name, once again, couldn't you get from the DataFields?

selvinsource commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/TreeModelUtils.scala:
```diff
@@ -0,0 +1,275 @@
+    val treeModel = new org.dmg.pmml.TreeModel()
+      .withModelName(modelName)
+      .withFunctionName(miningFunctionType)
+      .withNode(buildStub(mllibTreeModel.topNode))
```
We could add
.withSplitCharacteristic(SplitCharacteristic.BINARY_SPLIT)
for completeness, as far as I understand MLlib uses a binary split.

selvinsource commented:
As overall comment, to add to my specific comments above, I would say we should create the field names once (possibly in the data dictionary as data fields) and re-use these when creating mining fields or predicates. By doing so we ensure we are not creating field names that are not in the dictionary (due to potential bugs in the code for example).

In parallel I am updating my pmml validator project so that it validates the models exported from the Decision Tree exporter for both classification and regression. I will let you know when I have the results.

selvinsource commented:
Here some initial results from my tests (trying to load the exported xml into JPMML for evaluation).

**Node with no predicate**
If you look at 
https://github.com/selvinsource/spark-pmml-exporter-validator/tree/decision_trees_validator
and follow the Decision Tree (Classification) instructions
you get
[code]
java -jar target/spark-pmml-exporter-validator-1.1.0-SNAPSHOT-jar-with-dependencies.jar DecisionTreeClassificationModel
DecisionTreeClassificationModel selected
Exception in thread "main" org.jpmml.manager.InvalidFeatureException (at or around line 30): Node
	at org.jpmml.evaluator.TreeModelEvaluator.evaluateNode(TreeModelEvaluator.java:155)
	at org.jpmml.evaluator.TreeModelEvaluator.handleTrue(TreeModelEvaluator.java:186)
[/code]

The reason being that the node 32 has no predicate (see generated xml https://github.com/selvinsource/spark-pmml-exporter-validator/blob/decision_trees_validator/src/main/resources/exported_pmml_models/decisiontree_classification.xml). 

If you look at the documentation
http://www.dmg.org/v4-2-1/TreeModel.html#xsdGroup_PREDICATE, it says "Each Node has one PREDICATE; that may be a SimplePredicate, a SetPredicate, a CompoundPredicate, a True, or a False.".

In this case, node 32, I assume we need True:
    <Node id="32" score="0.0"><True/></Node>

This should apply to all other nodes with no predicate.

Once fixed, please try to run the code above and verify the evaluation is correct by running
    java -jar target/spark-pmml-exporter-validator-1.1.0-SNAPSHOT-jar-with-dependencies.jar DecisionTreeClassificationModel
You should get Class 0 for the first item and Class 1 for the second as the scala code (https://github.com/selvinsource/spark-pmml-exporter-validator/blob/decision_trees_validator/src/main/resources/spark_shell_exporter/decisiontree_breastcancerwisconsin.scala).

**Data Fields and Mining Fields**
I noticed that in order to evaluate the model in JPMML I need to provide only a subset of the fields and in this order: field_0, field_5, field_7, field_1. This is inconsistent with the scala code (Spark) whereas for prediction I still provide the features in the same order and number of fields of the training data.
It is not a big deal, but it would be nice to have a consistent behavior to the spark prediction. Also over time the exported decision tree can change the order of fields and the number of fields therefore the application using JPMML needs to be changed every time to match the number of fields and their order. It would be easier if the exported model has all the fields in the normal order field_0, field_1, field_2... so that as input we can still provide the full list of fields (however some will be ignored by the decision tree but that is fine). The advantage is that over time no code change is required to the decoupled application using JPMML and the same input vector of the one we use for training the model.
@mengxr What is your opinion on this point?

**Decision Tree Regression**
Once the above points have been addressed I will do some check on the regression model, if you look at https://github.com/selvinsource/spark-pmml-exporter-validator/blob/decision_trees_validator/src/main/java/org/selvinsource/spark_pmml_exporter_validator/SparkPMMLExporterValidator.java it is marked as TODO.

JasmineGeorge commented:
l will make the changes and get back. Thanks for all the comments.

JasmineGeorge committed:
changes to define fieldname only once, added treemodel split charecteristics as 'BinarySplit' and added default True predicate for leaf nodes

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41830 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41830/consoleFull) for   PR 7842 at commit [`0f769c9`](https://github.com/apache/spark/commit/0f769c94ef42de510b5b904b999eeb47c7de34f5).

SparkQA commented:
[Test build #41830 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41830/console) for   PR 7842 at commit [`0f769c9`](https://github.com/apache/spark/commit/0f769c94ef42de510b5b904b999eeb47c7de34f5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41830/
Test FAILed.

JasmineGeorge committed:
fixed stylecheck issues

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41832/consoleFull) for   PR 7842 at commit [`26dd880`](https://github.com/apache/spark/commit/26dd88022addfcb153ac4744519c4050f1852ddb).

SparkQA commented:
[Test build #41832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41832/console) for   PR 7842 at commit [`26dd880`](https://github.com/apache/spark/commit/26dd88022addfcb153ac4744519c4050f1852ddb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41832/
Test PASSed.

JasmineGeorge commented:
Made the changes. Now the FieldName is instantiated once and reused by DataField, MiningField and Predicates. Added default True predicate for leaf nodes and the Tree model SplitCharectistics as 'BinarySplit'.

JasmineGeorge commented:
Probably missed one point, that is returning the miningfields and datafields in normal order. Will do that tonight and checkin.

JasmineGeorge commented:
Had one question regarding returning the fields in normal order. By normal order you mean sorted lexicographically or by the double value of the fields e.g. field_1, field_11, field_3, you want the order to be field_1,field_11, field_3 or field_1,field_3, field_11

selvinsource commented:
As explained in details above, the order should be the same as the input vector (used for training the model).

If you look at https://github.com/selvinsource/spark-pmml-exporter-validator/blob/decision_trees_validator/src/main/resources/spark_shell_exporter/decisiontree_breastcancerwisconsin.scala
we take the first 9 values from the file as input vector:
```val data = sc.textFile("../datasets/breastcancerwisconsin_binaryclassification.csv")
val parsedData = data.map { line =>
  val parts = line.split(',')
  LabeledPoint(parts.last.toDouble, Vectors.dense(parts.take(9).map(_.toDouble)))
}```

after building the moded, the prediction uses the same number of fields and order of the input vector (5,1,1,1,2,1,3,1,1):
```val model = DecisionTree.trainClassifier(parsedData, numClasses, categoricalFeaturesInfo,
  impurity, maxDepth, maxBins)
var predictedValue = model.predict(Vectors.dense(5,1,1,1,2,1,3,1,1))```

So it would be nice if the exported tree listed the fields in the same order of the input vector.

As I said it is not a big deal, but it would be nice to have.

selvinsource commented:
Also, did you verify the evaluation is correct by running
java -jar target/spark-pmml-exporter-validator-1.1.0-SNAPSHOT-jar-with-dependencies.jar DecisionTreeClassificationModel?
In my opinion, **this is the most important test to ensure the exported model is correct**.
You should get Class 0 for the first item and Class 1 for the second as the scala code (https://github.com/selvinsource/spark-pmml-exporter-validator/blob/decision_trees_validator/src/main/resources/spark_shell_exporter/decisiontree_breastcancerwisconsin.scala).

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

JasmineGeorge commented:
found issues when ran the evaluator. Wait for new changes.

JasmineGeorge committed:
fixed bugs found after evaluation using jpmml evaluator

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42264 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42264/consoleFull) for   PR 7842 at commit [`e997621`](https://github.com/apache/spark/commit/e997621a1f840022b5e332805e859c748a1e47e8).

SparkQA commented:
[Test build #42264 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42264/console) for   PR 7842 at commit [`e997621`](https://github.com/apache/spark/commit/e997621a1f840022b5e332805e859c748a1e47e8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42264/
Test FAILed.

JasmineGeorge committed:
fixed unit test

AmplabJenkins commented:
Merged build triggered.

JasmineGeorge committed:
stylecheck issues resolved

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42266 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42266/consoleFull) for   PR 7842 at commit [`60ab25b`](https://github.com/apache/spark/commit/60ab25b237acf5ab4f46e4e8e53e7d5cdee8f3d1).

SparkQA commented:
[Test build #42266 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42266/console) for   PR 7842 at commit [`60ab25b`](https://github.com/apache/spark/commit/60ab25b237acf5ab4f46e4e8e53e7d5cdee8f3d1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42266/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42265/
Test FAILed.

JasmineGeorge committed:
style check issues resolved

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42268/consoleFull) for   PR 7842 at commit [`6912287`](https://github.com/apache/spark/commit/6912287119b9c95b6998388b99a696c535146bbc).

SparkQA commented:
[Test build #42268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42268/console) for   PR 7842 at commit [`6912287`](https://github.com/apache/spark/commit/6912287119b9c95b6998388b99a696c535146bbc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42268/
Test PASSed.

JasmineGeorge commented:
When I ran the evaluator I realized the evaluator expects that the root node predicate should always have predicate True.
And every node has the predicate condition to select that node.
Unlike MLLib where the split is for choosing left or right child node. For continuous feature  Split left if feature <= threshold, else right.
For categorical feature left if categorical feature value is in this set, else right.
I had to move the predicates to the left and right nodes appropriately in pmml, instead of the parent node which holds the split in MLLib.

Also in case of classification there must be class field in datafield and mining field which holds the predicted values.
Added those too.

About maintaining order of the MiningFields same as the input.
The input is a vector of Double and the features are essentially ordered
according to their index in the vector.

Made the changes to sort mining and datafields to be ordered according to the 
number part of the field names, which is their index in the input vector.

JasmineGeorge commented:
There are changes in the spark-pmml-exporter-validator. the pmml for decision-tree classification and also the tree evaluator method in SparkPMMLExporterValidator.java. I have done the required changes and issued a pull request in the spark-pmml-exporter-validator project

mengxr commented:
@selvinsource Could you help make another pass?

selvinsource commented:
Yeah, I am planning to review what @JasmineGeorge did soon, I have been a bit busy lately.

selvinsource commented:
Exported models (regression and classification trees) look good now: the validator generates the same results as spark.

Some minor comments:
- I would rename TreeModelUtils to something like PMMLTreeModelUtils...
- For the regression model we could add a target field for completeness (see GeneralizedLinearPMMLModelExport)
- For the classification model, I can see you generate a class field with values class_0 and class_1 but then the scored values are 0.0 and 1.0. We could align them to be consistent (either the label or the double value).

@mengxr If you open a Jira ticket and assign it to me, I can update the PMML documentation to include the new decision tree model.

JasmineGeorge committed:
TreeModelUtils renamed to PMMLTreeModelUtils, added target field for DT regression model for completeness, changed class names from class_<value> to keep the double value instead

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

JasmineGeorge committed:
updated classes datatype to double in DT with classification

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42752 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42752/consoleFull) for   PR 7842 at commit [`642a70e`](https://github.com/apache/spark/commit/642a70e13beec833cf3f3aecd6c5987e74355413).

SparkQA commented:
[Test build #42752 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42752/console) for   PR 7842 at commit [`642a70e`](https://github.com/apache/spark/commit/642a70e13beec833cf3f3aecd6c5987e74355413).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42752/
Test FAILed.

JasmineGeorge committed:
fixed stylecheck errors

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42754 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42754/consoleFull) for   PR 7842 at commit [`1fd4d77`](https://github.com/apache/spark/commit/1fd4d7708f48fc657202690ca2dc04cf1c1b78fe).

SparkQA commented:
[Test build #42754 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42754/console) for   PR 7842 at commit [`1fd4d77`](https://github.com/apache/spark/commit/1fd4d7708f48fc657202690ca2dc04cf1c1b78fe).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42754/
Test PASSed.

JasmineGeorge commented:
Changes Done 
- renamed TreeModelUtils to PMMLTreeModelUtils
- added target field for the regression model 
- for the classification model, now generates a class field with double values same as the scored values are 0.0 and 1.0 for consistency.

Also updated the pmml xmls in the spark-pmml-exporter-validator project.

selvinsource commented:
Thanks, looks good to me.

JasmineGeorge commented:
Thanks a lot @selvinsource.
@mengxr is there anything we should be adding/changing ? Please let me know.
Once these changes are merged I will start with the Random Forest and Gradient Boosted Trees.

JasmineGeorge commented:
@mengxr whats the next step ??

JasmineGeorge commented:
@coderxiang I was just wondering what would be the next step ?
Haven't heard from anyone for 3 weeks.

mengxr commented:
@JasmineGeorge Sorry for late response! We are running out of code review bandwidth. How about you and @yinxusen helping review each other's PR? His PR is at https://github.com/apache/spark/pull/9057.

@yinxusen Please let me know whether you can make another pass on this PR. Thanks!

yinxusen commented:
@mengxr Sure I'll reivew it.

JasmineGeorge commented:
Sure @mengxr I can review @yinxusen PR #9057

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExport.scala:
```diff
@@ -0,0 +1,48 @@
+  private def populateDecisionTreePMMLModel(model: DecisionTreeModel): Unit = {
+
+    pmml.getHeader.setDescription("decision tree")
+
```
No duplicated blank lines is better.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExport.scala:
```diff
@@ -0,0 +1,48 @@
+    val dataDictionary = new DataDictionary()
+      .withDataFields(dataFields.asJava)
+      .withNumberOfFields(dataFields.length)
+
```
ditto

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+package org.apache.spark.mllib.pmml.export
+
+import scala.collection.mutable
+import scala.collection.JavaConverters._
```
orders of import

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+
+  /** Build a pmml tree stub given the root mllib node. */
+  private def buildStub(
+              rootDTNode: Node,
```
four spaces before function definition. The same for the following line.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+    // for completeness
+    dtModel.algo match {
+
+      case Algo.Classification => {
```
The curly braces are not needed. The same for the following `case` statement.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+              rootDTNode: Node,
+              algo: Algo): (PMMLNode, List[MiningField], List[DataField], List[PMMLValue]) = {
+
+    val miningFields = MutableList[MiningField]()
```
mutable.MutableList is better.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+        val fieldName = FieldName.create(FieldNamePrefix + rootNode.split.get.feature)
+        val dataField = getDataField(rootNode, fieldName).get
+
+        if (!dataFields.get(dataField.getName.getValue).isDefined) {
```
I think you can use `dataFields.get(dataField.getName.getValue).isEmpty` instread of the `!xxx.isDefined`

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+            .withName(dataField.getName)
+            .withUsageType(FieldUsageType.ACTIVE)
+
+        } else if (dataField.getOpType != OpType.CONTINUOUS.value()) {
```
remove `.value()` in the end, because the type of both sides are not the same. `getOpType` gets the OpType, while `OpType.CONTINUOUS.value()` gets a string.

JasmineGeorge replied:
Good  catch. Thank You !

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+    if (dtField.getOpType == OpType.CATEGORICAL) {
+
+      val existingValues = dtField.getValues.asScala
+        .groupBy { case category => category.getValue }.toMap
```
remove `toMap` in the end, coz the groupBy returns a Map

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+        .groupBy { case category => category.getValue }.toMap
+
+      values.foreach(category => {
+        if (!existingValues.get(category.getValue).isDefined) {
```
use `isEmpty`

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+      val featureType = split.featureType
+
+      featureType match {
+        case FeatureType.Continuous => {
```
redundant curly braces

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/tree/model/DecisionTreeModel.scala:
```diff
@@ -45,7 +46,7 @@ import org.apache.spark.util.Utils
 class DecisionTreeModel @Since("1.0.0") (
     @Since("1.0.0") val topNode: Node,
-    @Since("1.0.0") val algo: Algo) extends Serializable with Saveable {
+    @Since("1.0.0") val algo: Algo) extends Serializable with Saveable with PMMLExportable{
```
a blank space before the final curly brace

yinxusen commented on mllib/src/test/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExportSuite.scala:
```diff
@@ -0,0 +1,431 @@
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+
+class DecisionTreePMMLModelExportSuite extends SparkFunSuite
+with MLlibTestSparkContext
```
I am not sure of the position of with statement. @mengxr Should we leave 2 or 4 spaces in this scenario?

yinxusen commented on mllib/src/test/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExportSuite.scala:
```diff
@@ -0,0 +1,431 @@
+    assert(pmmlRootNode.getNodes != null && pmmlRootNode.getNodes.size() == 2)
+    assert(pmmlRootNode.getId === "1")
+    // validate the root node predicate is populated as expected
+    val predicate = pmmlRootNode.getPredicate()
```
remove `()` in the end.

yinxusen commented on mllib/src/test/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExportSuite.scala:
```diff
@@ -0,0 +1,431 @@
+    val predicateR1 = pmmlRightNode_L1.getPredicate
+    assert(predicateR1 != null)
+    assert(predicateR1.isInstanceOf[True])
+
```
too many empty lines.

yinxusen commented on mllib/src/test/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExportSuite.scala:
```diff
@@ -0,0 +1,431 @@
+    assert(pmmlLeftNode_L2.getId === "4")
+    assert(pmmlLeftNode_L2.getScore == "1.0")
+    // validate predicate for level 2 split is populated as expected
+    val predicateL2 = pmmlLeftNode_L2.getPredicate()
```
remove `()` in the end

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLModelExportFactory.scala:
```diff
@@ -55,6 +53,8 @@ private[mllib] object PMMLModelExportFactory {
           throw new IllegalArgumentException(
             "PMML Export not supported for Multinomial Logistic Regression")
         }
+      case decisionTreeModel : DecisionTreeModel =>
```
remove the empty blank before `:`

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+  }
+
+  /** Get PMML mining function type for given mlLib Algo. */
+  private def getPMMLMiningFunctionType(mlLibAlgo: Algo): MiningFunctionType = {
```
There is no need to create a function here, coz it is only a match .. case statement.

You may use this:

```scala
dtModel.algo match {
  case ... =>
  case ... =>
}
```

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+    var (rootNode, miningFields, dataFields, classes) = buildStub(dtModel.topNode, dtModel.algo)
+
+    // adding predicted classes for classification and target field for regression
+    // for completeness
```
You can merge the two lines of comment.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+    }
+
+    val miningSchema = new MiningSchema()
+      .withMiningFields(miningFields.asJava)
```
It's better to merge the line with the previous one for compactness.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+      .withMiningFields(miningFields.asJava)
+
+    treeModel.withNode(rootNode)
+      .withMiningSchema(miningSchema)
```
same here

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+      .withMiningSchema(miningSchema)
+
+    (treeModel, dataFields)
+
```
remove the blank line.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+      val rootPMMLNode = createNode(rootNode)
+      rootPMMLNode.withPredicate(predicate)
+
+      var leftPredicate: Predicate = new True()
```
There is no need to use `var` here, coz the `leftPredicate` and `rightPredicate` are both assigned with value once.

JasmineGeorge replied:
Actually they are being assigned value once more inside the if block
if (rootNode.split.isDefined) {

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+
+        leftPredicate = getPredicate(rootNode, Some(dataField.getName), true)
+        rightPredicate = getPredicate(rootNode, Some(dataField.getName), false)
+
```
remove the empty line.

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+      pmmlValues)
+
+    result
+
```
remove the blank line

yinxusen commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,282 @@
+
+  /** Create equivalent PMML node for given mlLib node. */
+  private def createNode(mlLibNode: Node): PMMLNode = {
+    val node = new PMMLNode()
```
there is no need to define the `node`, new the PMMLNode

yinxusen commented on mllib/src/test/scala/org/apache/spark/mllib/pmml/export/DecisionTreePMMLModelExportSuite.scala:
```diff
@@ -0,0 +1,431 @@
+    assert(predicate1.asInstanceOf[SimplePredicate].getField.getValue === "field_100")
+    assert(predicate1.asInstanceOf[SimplePredicate].getValue === "10.0")
+    assert(predicate1.asInstanceOf[SimplePredicate].getOperator == SimplePredicate.Operator
+          .LESS_OR_EQUAL)
```
the indent should be 2 spaces.

JasmineGeorge committed:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

JasmineGeorge committed:
bug fixes and formatting fixes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44532 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44532/consoleFull)** for PR 7842 at commit [`1214601`](https://github.com/apache/spark/commit/121460162a5bf56d99dd4a9fa375863a9164aa99).

SparkQA commented:
**[Test build #44532 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44532/consoleFull)** for PR 7842 at commit [`1214601`](https://github.com/apache/spark/commit/121460162a5bf56d99dd4a9fa375863a9164aa99).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44532/
Test FAILed.

JasmineGeorge committed:
stylecheck fixes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44533/consoleFull)** for PR 7842 at commit [`8390211`](https://github.com/apache/spark/commit/839021103adfda5a6f17008b877728bf527169f3).

SparkQA commented:
**[Test build #44533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44533/consoleFull)** for PR 7842 at commit [`8390211`](https://github.com/apache/spark/commit/839021103adfda5a6f17008b877728bf527169f3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44533/
Test PASSed.

JasmineGeorge commented:
@yinxusen does everything look good now ?

yinxusen commented:
@mengxr @JasmineGeorge LGTM

holdenk commented:
Since @yinxusen gave this an looks good, maybe @jkbradley the umbrela issue reporter could take a look. @JasmineGeorge while this still merges after all these months it might be good to update from master just to make sure everything is still good (we've got some new style tests as well). While this is in the old MLLib code base it could be used as a basis for similar export work in the new ML code base.

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

SparkQA commented:
**[Test build #56311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56311/consoleFull)** for PR 7842 at commit [`cc1ee1f`](https://github.com/apache/spark/commit/cc1ee1f6f19b70c4fd06cbd25fcfa8fed055bed6).

SparkQA commented:
**[Test build #56311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56311/consoleFull)** for PR 7842 at commit [`cc1ee1f`](https://github.com/apache/spark/commit/cc1ee1f6f19b70c4fd06cbd25fcfa8fed055bed6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56311/
Test FAILed.

Commit:
fixed style issues

SparkQA commented:
**[Test build #56312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56312/consoleFull)** for PR 7842 at commit [`21e3616`](https://github.com/apache/spark/commit/21e3616595cea86d51478bf7651cfc8c5070f836).

SparkQA commented:
**[Test build #56312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56312/consoleFull)** for PR 7842 at commit [`21e3616`](https://github.com/apache/spark/commit/21e3616595cea86d51478bf7651cfc8c5070f836).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56312/
Test FAILed.

Commit:
fixed style issues

SparkQA commented:
**[Test build #56393 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56393/consoleFull)** for PR 7842 at commit [`49d0b80`](https://github.com/apache/spark/commit/49d0b804dc67870da11cef6ae957eccac53762d1).

SparkQA commented:
**[Test build #56393 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56393/consoleFull)** for PR 7842 at commit [`49d0b80`](https://github.com/apache/spark/commit/49d0b804dc67870da11cef6ae957eccac53762d1).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56393/
Test FAILed.

Commit:
Fixed compilation issues related to jpmml apis changes

SparkQA commented:
**[Test build #56419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56419/consoleFull)** for PR 7842 at commit [`7330966`](https://github.com/apache/spark/commit/733096684c9cd7ffefd5082c523b8c54d1807df4).

SparkQA commented:
**[Test build #56419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56419/consoleFull)** for PR 7842 at commit [`7330966`](https://github.com/apache/spark/commit/733096684c9cd7ffefd5082c523b8c54d1807df4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56419/
Test FAILed.

Commit:
fixed style issues

SparkQA commented:
**[Test build #56423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56423/consoleFull)** for PR 7842 at commit [`b80ee0d`](https://github.com/apache/spark/commit/b80ee0d8f48cb9e21239f25dcff4d93b55715e6d).

SparkQA commented:
**[Test build #56423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56423/consoleFull)** for PR 7842 at commit [`b80ee0d`](https://github.com/apache/spark/commit/b80ee0d8f48cb9e21239f25dcff4d93b55715e6d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56423/
Test PASSed.

JasmineGeorge commented:
ready for merge.
Which is the new mllib code base ?? mllib-local ?

jkbradley commented:
mllib-local is only for local linear algebra currently.  You can use the spark.ml package as the new one.  My apologies, but I don't think this will make 2.0 since we need to focus on bug/API fixes.  I do want to get this merged before too long though.

skonto commented:
@coderxiang what is the plan for this?

jomach commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,261 @@
+import scala.collection.JavaConverters._
+
+import org.dmg.pmml.{Node => PMMLNode, Value => PMMLValue, _}
+
```
remove blank Line

JasmineGeorge replied:
when i remove these lines scala styletest fails. It requires blank lines between different groups

jomach commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    val pmmlTreeRootNode = buildStubInternal(rootDTNode, new True())
+
```
remove blank Line

jomach commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,261 @@
+    val pmmlTreeRootNode = buildStubInternal(rootDTNode, new True())
+
+    val pmmlValues = classes.toList.distinct.map(doubleVal => new PMMLValue(doubleVal.toString))
+
```
remove blank Line

jomach commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala:
```diff
@@ -0,0 +1,261 @@
+
+      require(fieldName.isDefined, "fieldName should not be None, it should be defined.")
+      val field = fieldName.get
+
```
remove blank Line

Commit:
Merge branch 'master' into SPARK-8545-PMML-improvement-umbrella

Commit:
removed unneccesary blank lines

SparkQA commented:
**[Test build #82995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82995/testReport)** for PR 7842 at commit [`85a4c32`](https://github.com/apache/spark/commit/85a4c326c5a573245e40d636c951f0824b7870bc).

SparkQA commented:
**[Test build #82995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82995/testReport)** for PR 7842 at commit [`85a4c32`](https://github.com/apache/spark/commit/85a4c326c5a573245e40d636c951f0824b7870bc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82995/
Test PASSed.

JasmineGeorge commented:
removed all blank lines except the one in the import statements between different groups. Tests have passed. Are we ready to merge ??

holdenk commented:
CC @MLNick who I know has been doing some related work.

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins OK to test.
cc @MLnick

holdenk commented:
Is this still a PR your working on? We've recently started adding PMML support to Spark ML and it would be cool to have this in the ML side instead of (or in addition to) the legacy Spark MLlib.

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins OK to test

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?
